### DISCUSSION TITLE
[QUESTION]I cannot figure out TE(transformer_engine)

### DISCUSSION THREAD

**Discussion by Guodanding** [NVIDIA/Megatron-LM#1077] [Q&A]
I have not installed nvidia transformer_engine. I try running some example scripts (e.g train_mixtral_8x7b_distributed.sh) but fail with the TE error:
```
[rank0]: Traceback (most recent call last):
[rank0]:   File "/share/home/wenjunyi/guokb/workspace/imple-moe/Megatron-LM/pretrain_gpt.py", line 246, in <module>
[rank0]:     pretrain(
[rank0]:   File "/share/home/wenjunyi/guokb/workspace/imple-moe/Megatron-LM/megatron/training/training.py", line 270, in pretrain
[rank0]:     model, optimizer, opt_param_scheduler = setup_model_and_optimizer(
[rank0]:   File "/share/home/wenjunyi/guokb/workspace/imple-moe/Megatron-LM/megatron/training/training.py", line 581, in setup_model_and_optimizer
[rank0]:     model = get_model(model_provider_func, model_type)
[rank0]:   File "/share/home/wenjunyi/guokb/workspace/imple-moe/Megatron-LM/megatron/training/training.py", line 455, in get_model
[rank0]:     model = model_provider_func(
[rank0]:   File "/share/home/wenjunyi/guokb/workspace/imple-moe/Megatron-LM/pretrain_gpt.py", line 74, in model_provider
[rank0]:     transformer_layer_spec = get_gpt_layer_with_transformer_engine_spec(args.num_experts, args.moe_grouped_gemm, args.qk_layernorm)
[rank0]:   File "/share/home/wenjunyi/guokb/workspace/imple-moe/Megatron-LM/megatron/core/models/gpt/gpt_layer_specs.py", line 63, in get_gpt_layer_with_transformer_engine_spec
[rank0]:     mlp = _get_mlp_module_spec(
[rank0]:   File "/share/home/wenjunyi/guokb/workspace/imple-moe/Megatron-LM/megatron/core/models/gpt/gpt_layer_specs.py", line 154, in _get_mlp_module_spec
[rank0]:     linear_fc1 = TEColumnParallelGroupedLinear
[rank0]: NameError: name 'TEColumnParallelGroupedLinear' is not defined
```
I have read **Megatron-LM/megatron/core/models/gpt/gpt_layer_specs.py** and find that _**HAVE_TE**_ has been set to False:
```
try:
    from megatron.core.transformer.custom_layers.transformer_engine import (
        TEColumnParallelGroupedLinear,
        TEDotProductAttention,
        TELayerNormColumnParallelLinear,
        TENorm,
        TERowParallelGroupedLinear,
        TERowParallelLinear,
    )

    HAVE_TE = True
except ImportError:

    HAVE_TE = False
```
But it is useless because the code is judged by:
`use_te = args.transformer_impl == "transformer_engine"`
or
`if args.transformer_impl == "transformer_engine" then`
I don't know how to set **_args.transformer_impl_** and what choices can be set.
Or Megatron-LM must be run with nvidia transformer_engine?

**Comment by clarence-lee-sheng** [comment]
You can set a --transformer-impl local flag

There is a --transformer-impl argument in megatron/training/arguments.py L707
(https://github.com/NVIDIA/Megatron-LM/blob/b76a7d32f7af8cd6441f957bb4ba80afb0db61ba/megatron/training/arguments.py#L707)

**Comment by Guodanding** [comment]
> You can set a --transformer-impl local flag
> 
> There is a --transformer-impl argument in megatron/training/arguments.py L707 (
> 
> [Megatron-LM/megatron/training/arguments.py](https://github.com/NVIDIA/Megatron-LM/blob/b76a7d32f7af8cd6441f957bb4ba80afb0db61ba/megatron/training/arguments.py#L707)
> 
> Line 707 in [b76a7d3](/NVIDIA/Megatron-LM/commit/b76a7d32f7af8cd6441f957bb4ba80afb0db61ba)
> 
>  group.add_argument('--transformer-impl', default='transformer_engine', 
> )

Thanks! 
I know i can set default to '**local**' to achieve my goal. But is there 'standard' way to set the args? I have try this in **train_mixtral_8x7b_distributed.sh**:
```
ENGINE_ARGS=(
    --transformer_impl local \
)

torchrun ${DISTRIBUTED_ARGS[@]} /share/home/wenjunyi/guokb/workspace/imple-moe/Megatron-LM/pretrain_gpt.py \
    ${MODEL_ARGS[@]} \
    ${MOE_ARGS[@]} \
    ${DATA_ARGS[@]} \
    ${TRAINING_ARGS[@]} \
    ${MODEL_PARALLEL_ARGS[@]} \
    ${LOGGING_ARGS[@]} \
    ${ENGINE_ARGS[@]}
```
but fail:
`pretrain_gpt.py: error: unrecognized arguments: --transformer_impl local`

**Reply by yuyangJin** [reply]
--transformer_impl => --transformer-impl

**Comment by clarence-lee-sheng** [comment]
For your arguments.py, is there the --transformer_impl argument?

**Comment by Guodanding** [comment]
> For your arguments.py, is there the --transformer_impl argument?

yes.

**Comment by 1195343015** [comment]
I have installed `transformer_engine`, but the same error.

**Comment by Guodanding** [comment]
Is Apex needed?

**Comment by 1195343015** [comment]
I think it's not necessary

**Comment by Guodanding** [comment]
ok.
In fact, after i set _**--transformer_impl argument**_ to **_local_**, i meet another error:
`[rank1]: AssertionError: (RMSNorm) is not supported in by torch Layernorm when instantiating WrappedTorchLayerNorm`
So i am wondeing if i miss some needed packages.

**Comment by 1195343015** [comment]
You'd better use PyTorch NGC container as your environment (which contains te and apex and so on),   but I confirm that there's some bugs with megatron here.

**Comment by Guodanding** [comment]
ok. Thanks.

**Comment by Guodanding** [comment]
>  I confirm that there's some bugs with megatron here.

agree.

**Comment by Guodanding** [comment]
After i install apex, the error changes to:
`[rank0]: AssertionError: (RMSNorm) is not supported in FusedLayerNorm when instantiating FusedLayerNorm`

**Reply by JavaZeroo** [reply]
Did you solve the problem?

**Comment by eliird** [comment]
Set the --normalization argument to LayerNorm. Local does not support RMSNorm implementation
